In [1]:
import pandas as pd
import numpy as np

# Hacker News.

Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

In [2]:
hn = pd.read_csv("hacker_news.csv")
pd.options.display.max_columns = 150 # to avoid truncated output 
hn.head(5)

id                                              title  \
0  12224879                          Interactive Dynamic Video   
1  10975351  How to Use Open Source and Shut the Fuck Up at...   
2  11964716  Florida DJs May Face Felony for April Fools' W...   
3  11919867       Technology ventures: From Idea to Enterprise   
4  10301696  Note by Note: The Making of Steinway L1037 (2007)   

                                                 url  num_points  \
0            http://www.interactivedynamicvideo.com/         386   
1  http://hueniverse.com/2016/01/26/how-to-use-op...          39   
2  http://www.thewire.com/entertainment/2013/04/f...           2   
3  https://www.amazon.com/Technology-Ventures-Ent...           3   
4  http://www.nytimes.com/2007/11/07/movies/07ste...           8   

   num_comments      author       created_at  
0            52    ne0phyte   8/4/2016 11:52  
1            10      josep2  1/26/2016 19:30  
2             1    vezycash  6/23/2016 22:20  
3             1     hswarna   6/17/2016 0:01  
4             2  walterbell   9/30/2015 4:12

The code above displays the first five rows of hackers news

In [3]:
import re

In [4]:
askpost = hn[hn["title"].str.contains("^Ask Hn", flags=re.I)]
askpost.reset_index(inplace=True)
askpost

index        id                                              title  url  \
0         7  12296411        Ask HN: How to improve my personal website?  NaN   
1        17  10610020  Ask HN: Am I the only one outraged by Twitter ...  NaN   
2        22  11610310  Ask HN: Aby recent changes to CSS that broke m...  NaN   
3        30  12210105    Ask HN: Looking for Employee #3 How do I do it?  NaN   
4        31  10394168  Ask HN: Someone offered to buy my browser exte...  NaN   
...     ...       ...                                                ...  ...   
1739  20039  10994357  Ask HN: Is it feasible to port Apple's Swift t...  NaN   
1740  20042  12241954     Ask HN: What to do when a developer goes dark?  NaN   
1741  20045  12029526                         Ask HN: Killer app for AR?  NaN   
1742  20048  11227969  Ask HN: How do you balance a serious relations...  NaN   
1743  20061  11743946    Ask HN: Why are papers still published as PDFs?  NaN   

      num_points  num_comments        author        created_at  
0              2             6  ahmedbaracat    8/16/2016 9:55  
1             28            29          tkfx  11/22/2015 13:43  
2              1             1     polskibus    5/2/2016 10:14  
3              1             3        sph130    8/2/2016 14:20  
4             28            17      roykolak  10/15/2015 16:38  
...          ...           ...           ...               ...  
1739           3            17      schappim    1/29/2016 9:42  
1740           3             3     bittysdad    8/7/2016 12:58  
1741           2             2     davidiach     7/4/2016 8:50  
1742          10             4        audace     3/5/2016 1:25  
1743           4             2         adius    5/21/2016 9:22  

[1744 rows x 8 columns]

Ask post shows rows that starts with ask post

In [5]:
hn.shape

(20100, 7)

In [6]:
showpost = hn[hn["title"].str.contains("^Show Hn", flags=re.I)]
showpost.reset_index(inplace=True)
showpost

index        id                                              title  \
0        13  10627194  Show HN: Wio Link  ESP8266 Based Web of Things...   
1        39  10646440                Show HN: Something pointless I made   
2        46  11590768  Show HN: Shanhu.io, a programming playground p...   
3        84  12178806  Show HN: Webscope  Easy way for web developers...   
4        97  10872799  Show HN: GeoScreenshot  Easily test Geo-IP bas...   
...     ...       ...                                                ...   
1157  19993  11222099  Show HN: Geocoding API built with government o...   
1158  19999  11735438  Show HN: Decorating: Animated pulsed for your ...   
1159  20014  10200913                           Show HN: Idea to startup   
1160  20065  11444393     Show HN: PhantomJsCloud, Headless Browser SaaS   
1161  20070  12149183  Show HN: Parse recipe ingredients using JavaSc...   

                                               url  num_points  num_comments  \
0                             https://iot.seeed.cc          26            22   
1                          http://dn.ht/picklecat/         747           102   
2                                https://shanhu.io           1             1   
3                           http://webscopeapp.com           3             3   
4                   https://www.geoscreenshot.com/           1             9   
...                                            ...         ...           ...   
1157                             https://latlon.io           6             6   
1158        https://github.com/ryukinix/decorating           3             1   
1159                     https://ideatostartup.org          14            17   
1160                    https://PhantomJsCloud.com           2             1   
1161  https://github.com/herkyl/ingredients-parser           6             2   

              author        created_at  
0            kfihihc  11/25/2015 14:03  
1            dhotson  11/29/2015 22:46  
2              h8liu   4/28/2016 18:05  
3          fastbrick    7/28/2016 7:11  
4         kpsychwave    1/9/2016 20:45  
...              ...               ...  
1157       evanmarks     3/4/2016 4:50  
1158           lerax    5/20/2016 3:48  
1159      nikhildaga   9/10/2015 22:17  
1160        novaleaf     4/7/2016 3:04  
1161  zongitsrinzler   7/23/2016 11:35  

[1162 rows x 8 columns]

Show post shows rows that starts with show post

In [7]:
other = hn.drop(hn[hn["title"].str.contains("^Ask Hn", flags=re.I)].index)
other.reset_index(inplace=True)
other

index        id                                              title  \
0          0  12224879                          Interactive Dynamic Video   
1          1  10975351  How to Use Open Source and Shut the Fuck Up at...   
2          2  11964716  Florida DJs May Face Felony for April Fools' W...   
3          3  11919867       Technology ventures: From Idea to Enterprise   
4          4  10301696  Note by Note: The Making of Steinway L1037 (2007)   
...      ...       ...                                                ...   
18351  20095  12379592  How Purism Avoids Intels Active Management Tec...   
18352  20096  10339284          YC Application Translated and Broken Down   
18353  20097  10824382  Microkernels are slow and Elvis didn't do no d...   
18354  20098  10739875                      How Product Hunt really works   
18355  20099  11680777  RoboBrowser: Your friendly neighborhood web sc...   

                                                     url  num_points  \
0                http://www.interactivedynamicvideo.com/         386   
1      http://hueniverse.com/2016/01/26/how-to-use-op...          39   
2      http://www.thewire.com/entertainment/2013/04/f...           2   
3      https://www.amazon.com/Technology-Ventures-Ent...           3   
4      http://www.nytimes.com/2007/11/07/movies/07ste...           8   
...                                                  ...         ...   
18351  https://puri.sm/philosophy/how-purism-avoids-i...          10   
18352  https://medium.com/@zreitano/the-yc-applicatio...           4   
18353  http://blog.darknedgy.net/technology/2016/01/0...         169   
18354  https://medium.com/@benjiwheeler/how-product-h...         695   
18355              https://github.com/jmcarp/robobrowser         182   

       num_comments         author        created_at  
0                52       ne0phyte    8/4/2016 11:52  
1                10         josep2   1/26/2016 19:30  
2                 1       vezycash   6/23/2016 22:20  
3                 1        hswarna    6/17/2016 0:01  
4                 2     walterbell    9/30/2015 4:12  
...             ...            ...               ...  
18351             6  AdmiralAsshat    8/29/2016 2:22  
18352             1       zreitano   10/6/2015 14:57  
18353           132    vezzy-fnord     1/2/2016 0:49  
18354           222          brw12  12/15/2015 19:32  
18355            58      pmoriarty    5/12/2016 1:43  

[18356 rows x 8 columns]

In [8]:
otherpost = other.drop(other[other["title"].str.contains("^Show Hn", flags=re.I)].index)
otherpost.reset_index(inplace=True)
otherpost

level_0  index        id  \
0            0      0  12224879   
1            1      1  10975351   
2            2      2  11964716   
3            3      3  11919867   
4            4      4  10301696   
...        ...    ...       ...   
17189    18351  20095  12379592   
17190    18352  20096  10339284   
17191    18353  20097  10824382   
17192    18354  20098  10739875   
17193    18355  20099  11680777   

                                                   title  \
0                              Interactive Dynamic Video   
1      How to Use Open Source and Shut the Fuck Up at...   
2      Florida DJs May Face Felony for April Fools' W...   
3           Technology ventures: From Idea to Enterprise   
4      Note by Note: The Making of Steinway L1037 (2007)   
...                                                  ...   
17189  How Purism Avoids Intels Active Management Tec...   
17190          YC Application Translated and Broken Down   
17191  Microkernels are slow and Elvis didn't do no d...   
17192                      How Product Hunt really works   
17193  RoboBrowser: Your friendly neighborhood web sc...   

                                                     url  num_points  \
0                http://www.interactivedynamicvideo.com/         386   
1      http://hueniverse.com/2016/01/26/how-to-use-op...          39   
2      http://www.thewire.com/entertainment/2013/04/f...           2   
3      https://www.amazon.com/Technology-Ventures-Ent...           3   
4      http://www.nytimes.com/2007/11/07/movies/07ste...           8   
...                                                  ...         ...   
17189  https://puri.sm/philosophy/how-purism-avoids-i...          10   
17190  https://medium.com/@zreitano/the-yc-applicatio...           4   
17191  http://blog.darknedgy.net/technology/2016/01/0...         169   
17192  https://medium.com/@benjiwheeler/how-product-h...         695   
17193              https://github.com/jmcarp/robobrowser         182   

       num_comments         author        created_at  
0                52       ne0phyte    8/4/2016 11:52  
1                10         josep2   1/26/2016 19:30  
2                 1       vezycash   6/23/2016 22:20  
3                 1        hswarna    6/17/2016 0:01  
4                 2     walterbell    9/30/2015 4:12  
...             ...            ...               ...  
17189             6  AdmiralAsshat    8/29/2016 2:22  
17190             1       zreitano   10/6/2015 14:57  
17191           132    vezzy-fnord     1/2/2016 0:49  
17192           222          brw12  12/15/2015 19:32  
17193            58      pmoriarty    5/12/2016 1:43  

[17194 rows x 9 columns]

Other post shows rows that did not starts with ask post and show post

In [9]:
askpost["num_comments"].dtype

dtype('int64')

In [10]:
ask_tot_comments = askpost["num_comments"].sum()
print(ask_tot_comments)
ask_avg_comments = askpost["num_comments"].mean()
print(ask_avg_comments)

24483
14.038417431192661


Ask post received a total comments of 24483 and the average comment for ask post is 14.038

In [11]:
show_tot_comments = showpost["num_comments"].sum()
print(show_tot_comments)
show_avg_comments = showpost["num_comments"].mean()
print(show_avg_comments)

11988
10.31669535283993


Show post received a total comments of 11988 and the average comment for show post is 10.31

In [12]:
import datetime as dt

In [13]:
cols = ["created_at", "num_comments"]
ask_result = askpost[cols]
ask_result

created_at  num_comments
0       8/16/2016 9:55             6
1     11/22/2015 13:43            29
2       5/2/2016 10:14             1
3       8/2/2016 14:20             3
4     10/15/2015 16:38            17
...                ...           ...
1739    1/29/2016 9:42            17
1740    8/7/2016 12:58             3
1741     7/4/2016 8:50             2
1742     3/5/2016 1:25             4
1743    5/21/2016 9:22             2

[1744 rows x 2 columns]

In [14]:
ask_result["created_at"] = pd.to_datetime(ask_result["created_at"]).dt.strftime('%H')
ask_result["created_at"]

C:\Users\HP USA\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


0       09
1       13
2       10
3       14
4       16
        ..
1739    09
1740    12
1741    08
1742    01
1743    09
Name: created_at, Length: 1744, dtype: object

In [15]:
ask_result

created_at  num_comments
0            09             6
1            13            29
2            10             1
3            14             3
4            16            17
...         ...           ...
1739         09            17
1740         12             3
1741         08             2
1742         01             4
1743         09             2

[1744 rows x 2 columns]

In [16]:
hour = dict(ask_result.pivot_table(index="created_at", values="num_comments", aggfunc=np.sum))
hour

{'num_comments': created_at
 00     447
 01     683
 02    1381
 03     421
 04     337
 05     464
 06     397
 07     267
 08     492
 09     251
 10     793
 11     641
 12     687
 13    1253
 14    1416
 15    4477
 16    1814
 17    1146
 18    1439
 19    1188
 20    1722
 21    1745
 22     479
 23     543
 Name: num_comments, dtype: int64}

This output shows the total number of comments ask post received per hour

In [17]:
avg_hour = dict(ask_result.pivot_table(index="created_at", values="num_comments", aggfunc=np.mean))
avg_hour

{'num_comments': created_at
 00     8.127273
 01    11.383333
 02    23.810345
 03     7.796296
 04     7.170213
 05    10.086957
 06     9.022727
 07     7.852941
 08    10.250000
 09     5.577778
 10    13.440678
 11    11.051724
 12     9.410959
 13    14.741176
 14    13.233645
 15    38.594828
 16    16.796296
 17    11.460000
 18    13.201835
 19    10.800000
 20    21.525000
 21    16.009174
 22     6.746479
 23     7.985294
 Name: num_comments, dtype: float64}

This output shows the average number of comments ask post received per hour

In [18]:
col = ["created_at", "num_comments"]
show_result = showpost[cols]
show_result

created_at  num_comments
0     11/25/2015 14:03            22
1     11/29/2015 22:46           102
2      4/28/2016 18:05             1
3       7/28/2016 7:11             3
4       1/9/2016 20:45             9
...                ...           ...
1157     3/4/2016 4:50             6
1158    5/20/2016 3:48             1
1159   9/10/2015 22:17            17
1160     4/7/2016 3:04             1
1161   7/23/2016 11:35             2

[1162 rows x 2 columns]

In [19]:
show_result["created_at"] = pd.to_datetime(show_result["created_at"]).dt.strftime('%H').copy()
show_result["created_at"]

C:\Users\HP USA\anaconda3\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


0       14
1       22
2       18
3       07
4       20
        ..
1157    04
1158    03
1159    22
1160    03
1161    11
Name: created_at, Length: 1162, dtype: object

In [20]:
show_avg_hour = dict(show_result.pivot_table(index="created_at", values="num_comments", aggfunc=np.mean))
show_avg_hour

{'num_comments': created_at
 00    15.709677
 01     8.785714
 02     4.233333
 03    10.629630
 04     9.500000
 05     3.052632
 06     8.875000
 07    11.500000
 08     4.852941
 09     9.700000
 10     8.250000
 11    11.159091
 12    11.803279
 13     9.555556
 14    13.441860
 15     8.102564
 16    11.655914
 17     9.795699
 18    15.770492
 19     9.800000
 20    10.200000
 21     5.787234
 22    12.391304
 23    12.416667
 Name: num_comments, dtype: float64}

# CONCLUSION

From this analysis we observed that averagely, ask post received more comments[14.038] than show post[10.31].
Based on this analysis, we conclude that to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00, seeing that ask post created at ths time received the highest average comments.